WebClient and Retrofit facade builders

TimurKasatkin opened the pull request.

TimurKasatkin committed:
WebClient and Retrofit facade builders

TimurKasatkin committed:
DynamicRequestOptionsClient API improved and NotRetryingHttpClientBuilder implemented

TimurKasatkin committed:
jfix-armeria-facade builder API enriched

- writeRequestTimeout for client with retries
- ability to specified predefined EndpointGroup

TimurKasatkin committed:
ProfiledConnectionPoolListenerTest stabilized

AnonymousVovus commented on buildSrc/src/main/kotlin/Dependencies.kt:
```diff
@@ -16,16 +16,17 @@ object Vers {
     const val dynamic_property = "2.0.7"
```
new version of dynamic-property has been released.  
dynamic-property-2.0.8

AnonymousVovus commented on jfix-armeria-facade/src/main/kotlin/ru/fix/armeria/facade/retrofit/RetrofitBuilderApi.kt:
```diff
@@ -0,0 +1,39 @@
+        profiler: Profiler,
+        shutdownTimeoutProp: DynamicProperty<Duration>
+    ): RetrofitHttpClientBuilder
+    fun setBlockingResponseReadingExecutor(
```
minor: please add empty lines between methods

TimurKasatkin replied:
I think it is OK to add empty lines here between methods enableNamedBlockingResponseReadingExecutor and setBlockingResponseReadingExecutor (all variations) due to their declaration length

But I think it is better to do not add empty lines in HttpClientBuilderApi. Motivation of this - visual grouping of polymorthic interchangeable methods. 
E.g.:
![image](https://user-images.githubusercontent.com/9992425/94460252-1e15e780-01c1-11eb-884e-1057cd4e6789.png)

AnonymousVovus commented on jfix-armeria-facade/src/main/kotlin/ru/fix/armeria/facade/Utils.kt:
```diff
@@ -0,0 +1,12 @@
+package ru.fix.armeria.facade
+
+internal sealed class Either<out A, out B> {
```
What is the motivation for this class? There are lot's of usages 22 Left and 19 Right. What is a purpuse of this?

TimurKasatkin replied:
It is a wrapper type, popular in functional programming (see https://www.scala-lang.org/api/2.9.3/scala/Either.html), used to declare, that some field can have strictly one of two different types. Without it we will need to declare something like:
```
class SomeObject private constructor(
    val optionA: Int? = null,
    val optionB: String? = null
) {
    constructor(optionA: Int): this(optionA, null)
    constructor(optionB: String): this(null, optionB)
}

//more complex and not type/null safe code on SomeObject's client side
```

Here it is mostly used for timeout properties, that can be either static Duration values, or dynamic DynamicProperty<Duration> values (not both at the same time). Motivation of such distinction - to do not force jfix-armeria-facade clients to depend on dynamic-property (so that they can choose by themselves).

TimurKasatkin replied:
we discussed it verbally and agreed, that Either is a good choice here

AnonymousVovus commented on jfix-armeria-facade/src/main/kotlin/ru/fix/armeria/facade/retrofit/impl/RetrofitBuilderImpl.kt:
```diff
@@ -0,0 +1,143 @@
+                            "$logPrefix closing blocking response reading executor..."
+                        }
+                        it.lazyExecutor.value.shutdown()
+                        val shutdownTimeout = when (val timeout = it.shutdownTimeout) {
```
looks like unnecessary complexity

TimurKasatkin replied:
relates to comment above

AnonymousVovus commented on jfix-armeria-facade/src/main/kotlin/ru/fix/armeria/facade/webclient/impl/BaseHttpClientBuilderImpl.kt:
```diff
@@ -0,0 +1,258 @@
+    internal val lazyClientName: () -> String
+        get() = baseBuilderState.clientNameCreator
+
+    override fun setClientName(clientName: String) = copyOfThisBuilder(
```
Why do we always need to create copy of builder and copy of its state?

TimurKasatkin replied:
Motivation here is to allow library clients to reuse some base builders so that they can declare sceleton builder and then specify per client settings. E.g.:

1. lib user could declare base builder with such features as retrying (common retry rule and max retry attempts), requests and connections profiling (common profiler object)
2. then lib user specify different client names (will affect metric names), endpoint groups, timeouts, etc.

TimurKasatkin committed:
jfix-armeria-facade now support features to do not force lib client to depend on all JFix components and also:

- aggregating-profiler 1.6.5 -> 1.6.6
- dynamic-property-api 2.0.7 -> 2.0.8
- jfix-stdlib 3.0.9 -> 3.0.12

- flat jfix-armeria-facade-all[-retrofit] are created for maven projects convenience (Maven doesn't support Gradle capabilities)

- HttpClient/RetrofitBuilderApi enriched with documentation, specifying what dependencies are required for this feature to work

TimurKasatkin committed:
armeria updated to 1.1.0

TimurKasatkin committed:
README enriched with information about jfix-armeria-facade AND HttpClient/RetrofitBuilder APIs' javadocs are enriched by info about corresponding Gradle capabilities.

AnonymousVovus reviewed (approved):
OK
